File: luau_decompiler_go/deserializer/deserializer.go
// Package deserializer converts raw Luau bytecode bytes into structured
// Bytecode/Proto/Instruction objects. Supports versions 3-7.
package deserializer

import (
	"fmt"

	bc "Geckocompiler/bytecode"
)

// decodeInstructionFields extracts A/B/C/D/E from a raw 32-bit instruction word.
func decodeInstructionFields(raw uint32) (op, a, b, c uint8, d int16, e int32) {
	op = uint8(raw & 0xFF)
	a = uint8((raw >> 8) & 0xFF)
	b = uint8((raw >> 16) & 0xFF)
	c = uint8((raw >> 24) & 0xFF)

	// D: signed 16-bit from bits 16-31
	dRaw := uint16((raw >> 16) & 0xFFFF)
	d = int16(dRaw)

	// E: signed 24-bit from bits 8-31
	eRaw := (raw >> 8) & 0xFFFFFF
	if eRaw >= 0x800000 {
		e = int32(eRaw) - 0x1000000
	} else {
		e = int32(eRaw)
	}
	return
}

// Deserialize parses a Luau bytecode buffer into a Bytecode structure.
func Deserialize(data []byte) (*bc.Bytecode, error) {
	reader := bc.NewReader(data)
	result := &bc.Bytecode{}

	// --- Header ---
	version, err := reader.ReadByte()
	if err != nil {
		return nil, fmt.Errorf("reading version: %w", err)
	}
	result.Version = int(version)
	if result.Version == 0 {
		return nil, fmt.Errorf("version 0 indicates a compilation error marker")
	}

	if result.Version >= 4 {
		tv, err := reader.ReadByte()
		if err != nil {
			return nil, fmt.Errorf("reading types version: %w", err)
		}
		result.TypesVersion = int(tv)
	}

	// --- String Table ---
	stringCount, err := reader.ReadVarint()
	if err != nil {
		return nil, fmt.Errorf("reading string count: %w", err)
	}
	result.Strings = make([]string, 0, stringCount)
	for i := 0; i < stringCount; i++ {
		strLen, err := reader.ReadVarint()
		if err != nil {
			return nil, fmt.Errorf("reading string %d length: %w", i, err)
		}
		s, err := reader.ReadString(strLen)
		if err != nil {
			return nil, fmt.Errorf("reading string %d: %w", i, err)
		}
		result.Strings = append(result.Strings, s)
	}

	// --- Proto Table ---
	if result.Version >= 7 {
		// Version 7 has an extra byte before proto count
		if _, err := reader.ReadByte(); err != nil {
			return nil, fmt.Errorf("reading v7 extra byte: %w", err)
		}
	}

	protoCount, err := reader.ReadVarint()
	if err != nil {
		return nil, fmt.Errorf("reading proto count: %w", err)
	}

	result.Protos = make([]*bc.Proto, 0, protoCount)
	for pIdx := 0; pIdx < protoCount; pIdx++ {
		proto, err := deserializeProto(reader, result, pIdx)
		if err != nil {
			return nil, fmt.Errorf("proto %d: %w", pIdx, err)
		}
		result.Protos = append(result.Protos, proto)
	}

	// --- Main proto index ---
	mainID, err := reader.ReadVarint()
	if err != nil {
		return nil, fmt.Errorf("reading main proto id: %w", err)
	}
	result.MainProtoID = mainID

	return result, nil
}

func deserializeProto(reader *bc.BytecodeReader, result *bc.Bytecode, pIdx int) (*bc.Proto, error) {
	proto := &bc.Proto{ProtoID: pIdx}

	maxStack, err := reader.ReadByte()
	if err != nil {
		return nil, fmt.Errorf("reading max_stack: %w", err)
	}
	proto.MaxStackSize = int(maxStack)

	numParams, err := reader.ReadByte()
	if err != nil {
		return nil, err
	}
	proto.NumParams = int(numParams)

	numUpvalues, err := reader.ReadByte()
	if err != nil {
		return nil, err
	}
	proto.NumUpvalues = int(numUpvalues)

	isVararg, err := reader.ReadByte()
	if err != nil {
		return nil, err
	}
	proto.IsVararg = isVararg != 0

	if result.Version >= 4 {
		flags, err := reader.ReadByte()
		if err != nil {
			return nil, err
		}
		proto.Flags = int(flags)

		typeInfoSize, err := reader.ReadVarint()
		if err != nil {
			return nil, err
		}
		if typeInfoSize > 0 {
			ti, err := reader.ReadBytes(typeInfoSize)
			if err != nil {
				return nil, err
			}
			proto.TypeInfo = ti
		}
	}

	// --- Instructions ---
	if err := deserializeInstructions(reader, proto); err != nil {
		return nil, fmt.Errorf("instructions: %w", err)
	}

	// --- Constants ---
	if err := deserializeConstants(reader, result, proto, pIdx); err != nil {
		return nil, fmt.Errorf("constants: %w", err)
	}

	// --- Child protos ---
	childCount, err := reader.ReadVarint()
	if err != nil {
		return nil, fmt.Errorf("reading child count: %w", err)
	}
	proto.ChildProtos = make([]int, 0, childCount)
	for i := 0; i < childCount; i++ {
		cid, err := reader.ReadVarint()
		if err != nil {
			return nil, err
		}
		proto.ChildProtos = append(proto.ChildProtos, cid)
	}

	// --- Debug info ---
	deserializeDebugInfo(reader, result, proto)

	return proto, nil
}

func deserializeInstructions(reader *bc.BytecodeReader, proto *bc.Proto) error {
	codeSize, err := reader.ReadVarint()
	if err != nil {
		return err
	}

	rawInsts := make([]uint32, 0, codeSize)
	for i := 0; i < codeSize; i++ {
		raw, err := reader.ReadUint32()
		if err != nil {
			return err
		}
		rawInsts = append(rawInsts, raw)
	}

	pc := 0
	for pc < len(rawInsts) {
		raw := rawInsts[pc]
		opRaw, a, b, c, d, e := decodeInstructionFields(raw)

		// Opcode decryption: (encryptedOp * 203) & 0xFF
		op := bc.LuauOpcode((uint16(opRaw) * 203) & 0xFF)

		inst := &bc.Instruction{
			PC:     pc,
			Opcode: op,
			OpName: bc.OpcodeName(op),
			A:      int(a),
			B:      int(b),
			C:      int(c),
			D:      int(d),
			E:      int(e),
			Aux:    -1,
			Raw:    raw,
		}

		// Check for AUX word
		if bc.OpcodesWithAux[op] && pc+1 < len(rawInsts) {
			inst.Aux = int(rawInsts[pc+1])
			pc += 2
		} else {
			pc++
		}

		proto.Instructions = append(proto.Instructions, inst)
	}
	return nil
}

func deserializeConstants(reader *bc.BytecodeReader, result *bc.Bytecode, proto *bc.Proto, pIdx int) error {
	constCount, err := reader.ReadVarint()
	if err != nil {
		return err
	}

	proto.Constants = make([]*bc.Constant, 0, constCount)
	for kIdx := 0; kIdx < constCount; kIdx++ {
		constType, err := reader.ReadByte()
		if err != nil {
			return err
		}
		ct := bc.LuauConstantType(constType)
		c := &bc.Constant{Index: kIdx, Type: ct}

		switch ct {
		case bc.ConstNil:
			c.Value = nil
		case bc.ConstBoolean:
			bv, err := reader.ReadByte()
			if err != nil {
				return err
			}
			c.Value = bv != 0
		case bc.ConstNumber:
			v, err := reader.ReadFloat64()
			if err != nil {
				return err
			}
			c.Value = v
		case bc.ConstString:
			strIdx, err := reader.ReadVarint()
			if err != nil {
				return err
			}
			c.Value = result.GetString(strIdx)
		case bc.ConstImport:
			v, err := reader.ReadUint32()
			if err != nil {
				return err
			}
			c.Value = v
		case bc.ConstTable:
			tableSize, err := reader.ReadVarint()
			if err != nil {
				return err
			}
			keys := make([]int, 0, tableSize)
			for j := 0; j < tableSize; j++ {
				k, err := reader.ReadVarint()
				if err != nil {
					return err
				}
				keys = append(keys, k)
			}
			c.Value = keys
		case bc.ConstClosure:
			v, err := reader.ReadVarint()
			if err != nil {
				return err
			}
			c.Value = v
		case bc.ConstVector:
			x, err := reader.ReadFloat32()
			if err != nil {
				return err
			}
			y, err := reader.ReadFloat32()
			if err != nil {
				return err
			}
			z, err := reader.ReadFloat32()
			if err != nil {
				return err
			}
			w, err := reader.ReadFloat32()
			if err != nil {
				return err
			}
			c.Value = [4]float32{x, y, z, w}
		default:
			fmt.Printf("Warning: Unknown constant type %d at proto %d, constant %d\n", constType, pIdx, kIdx)
			proto.Constants = append(proto.Constants, c)
			return nil // stop reading constants for this proto
		}

		proto.Constants = append(proto.Constants, c)
	}
	return nil
}

func deserializeDebugInfo(reader *bc.BytecodeReader, result *bc.Bytecode, proto *bc.Proto) {
	// Wrapped in recovery since debug info may be truncated
	defer func() { recover() }()

	if result.Version >= 7 {
		lineDefined, err := reader.ReadVarint()
		if err != nil {
			return
		}
		proto.LineDefined = lineDefined

		debugNameIdx, err := reader.ReadVarint()
		if err != nil {
			return
		}
		proto.DebugName = result.GetString(debugNameIdx)
	}

	hasLineInfo, err := reader.ReadByte()
	if err != nil {
		return
	}
	if hasLineInfo != 0 {
		parseLineInfo(reader, proto)
	}

	hasDebugInfo, err := reader.ReadByte()
	if err != nil {
		return
	}
	if hasDebugInfo != 0 {
		parseLocalVarInfo(reader, result, proto)
	}
}

func parseLineInfo(reader *bc.BytecodeReader, proto *bc.Proto) {
	lineGapLog2, err := reader.ReadByte()
	if err != nil {
		return
	}
	lineGap := 1 << lineGapLog2
	numInstructions := len(proto.Instructions)

	// Line deltas — one per raw instruction (including AUX words)
	codeSize := numInstructions
	// Recalculate code size: each instruction with AUX counts as 2 raw slots
	rawCodeSize := 0
	for _, inst := range proto.Instructions {
		rawCodeSize++
		if inst.HasAux() {
			rawCodeSize++
		}
	}
	codeSize = rawCodeSize

	lineInfo := make([]int, 0, codeSize)
	for i := 0; i < codeSize; i++ {
		b, err := reader.ReadByte()
		if err != nil {
			return
		}
		delta := int(int8(b))
		lineInfo = append(lineInfo, delta)
	}

	numIntervals := 0
	if codeSize > 0 {
		numIntervals = ((codeSize - 1) / lineGap) + 1
	}

	absLines := make([]int32, 0, numIntervals)
	for i := 0; i < numIntervals; i++ {
		v, err := reader.ReadInt32()
		if err != nil {
			return
		}
		absLines = append(absLines, v)
	}

	if len(absLines) > 0 {
		currentLine := int(absLines[0])
		resolved := make([]int, 0, codeSize)
		for i := 0; i < codeSize; i++ {
			intervalIdx := i / lineGap
			if i%lineGap == 0 && intervalIdx < len(absLines) {
				currentLine = int(absLines[intervalIdx])
			} else {
				currentLine += lineInfo[i]
			}
			resolved = append(resolved, currentLine)
		}
		proto.LineInfo = resolved
	}
}

func parseLocalVarInfo(reader *bc.BytecodeReader, result *bc.Bytecode, proto *bc.Proto) {
	localCount, err := reader.ReadVarint()
	if err != nil {
		return
	}
	for i := 0; i < localCount; i++ {
		nameIdx, err := reader.ReadVarint()
		if err != nil {
			return
		}
		startPC, err := reader.ReadVarint()
		if err != nil {
			return
		}
		endPC, err := reader.ReadVarint()
		if err != nil {
			return
		}
		reg, err := reader.ReadByte()
		if err != nil {
			return
		}
		proto.LocalVars = append(proto.LocalVars, bc.LocalVarInfo{
			Name:    result.GetString(nameIdx),
			StartPC: startPC,
			EndPC:   endPC,
			Reg:     int(reg),
		})
	}

	upvalCount, err := reader.ReadVarint()
	if err != nil {
		return
	}
	for i := 0; i < upvalCount; i++ {
		nameIdx, err := reader.ReadVarint()
		if err != nil {
			return
		}
		proto.UpvalueNames = append(proto.UpvalueNames, result.GetString(nameIdx))
	}
}

File: luau_decompiler/deserializer.py
"""
Luau Bytecode Deserializer
Parses raw bytecode bytes into structured Proto/Constant/Instruction objects.
Supports Luau bytecode versions 3-7.
"""

from .utils import BytecodeReader, decode_import_id
from .bytecode import Bytecode, Proto, Constant, Instruction
from .enums import LuauOpcode, LuauConstantType, OPCODES_WITH_AUX, OPCODE_FORMAT
import struct


def _decode_instruction(raw: int) -> dict:
    """Decode a raw 32-bit instruction word into its fields."""
    op = raw & 0xFF
    a = (raw >> 8) & 0xFF
    b = (raw >> 16) & 0xFF
    c = (raw >> 24) & 0xFF

    # D is a signed 16-bit value from bits 16-31
    d = (raw >> 16) & 0xFFFF
    if d >= 0x8000:
        d -= 0x10000

    # E is a signed 24-bit value from bits 8-31
    e = (raw >> 8) & 0xFFFFFF
    if e >= 0x800000:
        e -= 0x1000000

    return {"op": op, "a": a, "b": b, "c": c, "d": d, "e": e}


def _get_opname(opcode: int) -> str:
    """Get the human-readable name of an opcode."""
    try:
        return LuauOpcode(opcode).name
    except ValueError:
        return f"UNKNOWN_{opcode:02X}"


def deserialize(data: bytes) -> Bytecode:
    """
    Deserialize a Luau bytecode buffer into a Bytecode object.
    """
    reader = BytecodeReader(data)
    bc = Bytecode()

    # --- Header ---
    bc.version = reader.read_byte()
    if bc.version == 0:
        raise ValueError("Bytecode error: version 0 indicates a compilation error marker")

    if bc.version >= 4:
        bc.types_version = reader.read_byte()

    # --- String Table ---
    string_count = reader.read_varint()
    for _ in range(string_count):
        str_len = reader.read_varint()
        s = reader.read_string(str_len)
        bc.strings.append(s)

    # --- Proto Table ---
    if bc.version >= 7:
        # Version 7 has an extra byte (usually 0) before proto count
        reader.read_byte()

    proto_count = reader.read_varint()
    for proto_idx in range(proto_count):
        proto = Proto(proto_id=proto_idx)

        proto.max_stack_size = reader.read_byte()
        proto.num_params = reader.read_byte()
        proto.num_upvalues = reader.read_byte()
        proto.is_vararg = bool(reader.read_byte())

        if bc.version >= 4:
            proto.flags = reader.read_byte()
            typeinfo_size = reader.read_varint()
            if typeinfo_size > 0:
                proto.typeinfo = reader.read_bytes(typeinfo_size)

        # --- Instructions ---
        code_size = reader.read_varint()
        raw_instructions = []
        for _ in range(code_size):
            raw_instructions.append(reader.read_uint32())

        # Decode instructions, handling AUX words and opcode mapping
        pc = 0
        while pc < len(raw_instructions):
            raw = raw_instructions[pc]
            fields = _decode_instruction(raw)
            
            # Opcode mapping for this version/obfuscation
            # EncryptedOp = (Op * 227) % 256
            # DecryptedOp = (EncryptedOp * 203) % 256
            op = (fields["op"] * 203) & 0xFF

            inst = Instruction(
                pc=pc,
                opcode=op,
                opname=_get_opname(op),
                a=fields["a"],
                b=fields["b"],
                c=fields["c"],
                d=fields["d"],
                e=fields["e"],
                raw=raw,
            )

            # Check if this opcode has an AUX word
            try:
                has_aux = LuauOpcode(op) in OPCODES_WITH_AUX
            except ValueError:
                has_aux = False

            if has_aux and (pc + 1) < len(raw_instructions):
                inst.aux = raw_instructions[pc + 1]
                pc += 2
            else:
                pc += 1

            proto.instructions.append(inst)

        # --- Constants ---
        const_count = reader.read_varint()
        for k_idx in range(const_count):
            const_type = reader.read_byte()
            const = Constant(index=k_idx, type=const_type)

            if const_type == LuauConstantType.NIL:
                const.value = None
            elif const_type == LuauConstantType.BOOLEAN:
                const.value = bool(reader.read_byte())
            elif const_type == LuauConstantType.NUMBER:
                const.value = reader.read_double()
            elif const_type == LuauConstantType.STRING:
                str_idx = reader.read_varint()
                const.value = bc.get_string(str_idx)
            elif const_type == LuauConstantType.IMPORT:
                const.value = reader.read_uint32()  # Encoded import ID
            elif const_type == LuauConstantType.TABLE:
                table_size = reader.read_varint()
                keys = []
                for _ in range(table_size):
                    keys.append(reader.read_varint())
                const.value = keys  # List of constant indices for keys
            elif const_type == LuauConstantType.CLOSURE:
                const.value = reader.read_varint()  # Proto index
            elif const_type == LuauConstantType.VECTOR:
                x = reader.read_float()
                y = reader.read_float()
                z = reader.read_float()
                w = reader.read_float()
                const.value = (x, y, z, w)
            else:
                print(f"Warning: Unknown constant type {const_type} at proto {proto_idx}, constant {k_idx}")
                # Try to guess - if we can't, we might stay out of sync. 
                # But let's try to just break this proto's constant reading.
                break

            proto.constants.append(const)

        # --- Child protos ---
        child_count = reader.read_varint()
        for _ in range(child_count):
            proto.child_protos.append(reader.read_varint())

        # --- Debug info ---
        # V7: lineDefined(varint) + debugname_idx(varint) come before lineinfo
        try:
            if bc.version >= 7:
                proto.line_defined = reader.read_varint()
                debugname_idx = reader.read_varint()
                proto.debug_name = bc.get_string(debugname_idx)

            has_lineinfo = reader.read_byte()
            if has_lineinfo:
                linegaplog2 = reader.read_byte()
                line_gap = 1 << linegaplog2
                num_instructions = code_size
                lineinfo = []
                for _ in range(num_instructions):
                    delta = reader.read_byte()
                    if delta >= 128:
                        delta -= 256
                    lineinfo.append(delta)
                num_intervals = ((num_instructions - 1) // line_gap) + 1 if num_instructions > 0 else 0
                abs_lines = []
                for _ in range(num_intervals):
                    abs_lines.append(reader.read_int32())
                if abs_lines:
                    current_line = abs_lines[0]
                    resolved_lines = []
                    for i in range(num_instructions):
                        interval_idx = i // line_gap
                        if i % line_gap == 0 and interval_idx < len(abs_lines):
                            current_line = abs_lines[interval_idx]
                        else:
                            current_line += lineinfo[i]
                        resolved_lines.append(current_line)
                    proto.line_info = resolved_lines

            has_debuginfo = reader.read_byte()
            if has_debuginfo:
                local_count = reader.read_varint()
                for _ in range(local_count):
                    name_idx = reader.read_varint()
                    start_pc = reader.read_varint()
                    end_pc = reader.read_varint()
                    reg = reader.read_byte()
                    proto.local_vars.append({
                        "name": bc.get_string(name_idx),
                        "start_pc": start_pc,
                        "end_pc": end_pc,
                        "reg": reg,
                    })
                upval_count = reader.read_varint()
                for _ in range(upval_count):
                    name_idx = reader.read_varint()
                    proto.upvalue_names.append(bc.get_string(name_idx))
        except Exception as e:
            print(f"Warning: Failed to read debug info for proto {proto_idx}: {e}")

        bc.protos.append(proto)

    # --- Main proto index ---
    bc.main_proto_id = reader.read_varint()

    return bc
